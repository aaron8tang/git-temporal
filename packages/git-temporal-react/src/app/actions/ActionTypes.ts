export enum ActionTypes {
  REQUEST_COMMITS = 'REQUEST_COMMITS',
  RECEIVE_COMMITS = 'RECEIVE_COMMITS',
  SELECT_PATH = 'SELECT_PATH',
  INVALIDATE_PATH = 'INVALIDATE_PATH',
  HIGHLIGHT_COMMIT = 'HIGHLIGHT_COMMIT',
  VIEW_COMMITS = 'VIEW_COMMITS',
  VIEW_FILES = 'VIEW_FILES',
  ADD_AUTHOR_FILTER = 'ADD_AUTHOR_FILTER',
  REMOVE_AUTHOR_FILTER = 'REMOVE_AUTHOR_FILTER',
  REMOVE_ALL_AUTHOR_FILTERS = 'REMOVE_ALL_AUTHOR_FILTERS',
}
