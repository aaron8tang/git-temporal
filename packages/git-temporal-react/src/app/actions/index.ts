import { ActionTypes } from './ActionTypes';

export const selectPath = path => (dispatch, _getState) => {
  // if this comes from a rename, follow the most current name
  const actualPath = path.replace(/\{(.*)\s=>\s(.*)\}/g, '$2');

  dispatch(fetchCommitsIfNeeded(actualPath));
  return { selectedPath: actualPath, type: ActionTypes.SELECT_PATH };
};

export const invalidatePath = path => ({
  selectedPath: path,
  type: ActionTypes.INVALIDATE_PATH,
});

export const addAuthorFilter = authorName => ({
  authorName,
  type: ActionTypes.ADD_AUTHOR_FILTER,
});

export const removeAuthorFilter = authorName => ({
  authorName,
  type: ActionTypes.REMOVE_AUTHOR_FILTER,
});

export const removeAllAuthorFilters = () => ({
  type: ActionTypes.REMOVE_ALL_AUTHOR_FILTERS,
});

export const highlightCommit = commitId => ({
  commitId,
  type: ActionTypes.HIGHLIGHT_COMMIT,
});

export const viewCommits = () => ({
  type: ActionTypes.VIEW_COMMITS,
});

export const viewFiles = () => ({
  type: ActionTypes.VIEW_FILES,
});

export const requestCommits = path => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_COMMITS,
});

export const receiveCommits = (path, json) => ({
  selectedPath: path,
  commits: json.commits,
  type: ActionTypes.RECEIVE_COMMITS,
});

const fetchCommits = path => dispatch => {
  dispatch(requestCommits(path));
  const pathParam = path && path.trim().length > 0 ? `?path=${path}` : '';
  // TODO : replace this with serviceBaseUrl when it is in
  return fetch(`http://localhost:11966/git-temporal/history${pathParam}`)
    .then(response => response.json())
    .then(json => dispatch(receiveCommits(path, json)));
};

const shouldFetchCommits = (state, path) => {
  if (state.isFetching) {
    return false;
  }
  return (
    state.didInvalidate ||
    !state.commits ||
    state.commits.length <= 0 ||
    state.selectedPath !== path
  );
};

export const fetchCommitsIfNeeded = path => (dispatch, getState) => {
  if (shouldFetchCommits(getState(), path)) {
    return dispatch(fetchCommits(path));
  }
};
