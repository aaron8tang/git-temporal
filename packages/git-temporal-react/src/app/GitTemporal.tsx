import React, { Component } from 'react';
import { connect } from 'react-redux';
import { debug } from '@git-temporal/logger';

import { GitTemporalProps, DispatchProps, StateProps } from 'app/interfaces';
import { selectPath } from 'app/actions';
import { handleVscodeMessages } from 'app/actions/vscode';
import { getGitTemporalContainerState } from 'app/selectors';
import { style } from 'app/styles';

import Header from 'app/containers/Header';
import Stats from 'app/containers/Stats';
import Authors from 'app/containers/Authors';
import Files from 'app/containers/Files';
import Commits from 'app/containers/Commits';
import Timeplot from 'app/containers/Timeplot';
import DifferenceViewer from 'app/containers/DifferenceViewer';

import { SpinnerContainer } from 'app/components/SpinnerContainer';
import { HorizontalScroller } from 'app/components/HorizontalScroller';
import { TransitionVisible } from 'app/components/TransitionVisible';

const outerScrollStyle = {
  _extends: 'flexRows',
  position: 'relative',
  minHeight: '100%',
  height: 'initial',
};

const innerScrollStyle = {
  position: 'relative',
  display: 'flex',
  flexDirection: 'row',
  flexGrow: 1,
};

const scrollerIconStyle = {
  transform: 'scaleY(8)',
  top: '35%',
};

const transitionStyle = {
  _extends: 'flexColumns',
  flexGrow: 1,
};

export class GitTemporal extends Component<
  GitTemporalProps & DispatchProps & StateProps
> {
  componentDidMount() {
    const { path, dispatch } = this.props;
    dispatch(selectPath(path));
    handleVscodeMessages(dispatch);
  }

  componentWillUnmount() {
    debug('unmounting GitTemporal');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.path !== this.props.path) {
      const { dispatch, path } = nextProps;
      dispatch(selectPath(path));
    }
  }

  render() {
    const { isFetching, viewCommitsOrFiles } = this.props;
    debug('rendering GitTemporal');

    return (
      <div style={style('page')}>
        <SpinnerContainer isSpinning={isFetching}>
          <div style={style('flexColumns', { height: '100%' })}>
            <Header />
            <TransitionVisible
              isVisible={!isFetching}
              style={style(transitionStyle)}
            >
              <Stats />
              <div style={{ display: 'flex', flexGrow: 1 }}>
                <HorizontalScroller
                  style={style(outerScrollStyle)}
                  innerStyle={style(innerScrollStyle)}
                  iconStyle={style(scrollerIconStyle)}
                >
                  <Authors />
                  {viewCommitsOrFiles === 'files' ? <Files /> : <Commits />}
                  <DifferenceViewer />
                </HorizontalScroller>
              </div>
              <Timeplot />
            </TransitionVisible>
          </div>
        </SpinnerContainer>
      </div>
    );
  }
}

export default connect(getGitTemporalContainerState)(GitTemporal);
