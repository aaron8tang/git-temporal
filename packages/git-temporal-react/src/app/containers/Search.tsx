import React, { Component } from 'react';
import { connect } from 'react-redux';

import { DispatchProps } from 'app/interfaces';
import { getSearchContainerState } from 'app/selectors';
import { style } from 'app/styles';
import { setSearch } from 'app/actions';
import { SearchInput } from 'app/components/SearchInput';
import { Popup } from 'app/components/Popup';
import { Selectable } from 'app/components/Selectable';

import { debounce } from 'app/utilities/debounce';

interface SearchProps {
  search?: string;
}

interface SearchLocalState {
  searchHasFocus: boolean;
  selectedSuggestion: string;
}

const initialState = {
  searchHasFocus: false,
  selectedSuggestion: 'all',
};

const containerStyle = {
  width: 250,
  overflow: 'visible',
  position: 'relative',
};

const searchInputStyle = {
  width: '100%',
  borderRadius: 10,
  marginLeft: 10,
  marginTop: 0,
};

const popupStyle = {
  marginRight: 0,
};

const suggestedSearchPrefixes = ['author:', 'commit:', 'file:'];

export class Search extends Component<
  SearchProps & DispatchProps,
  SearchLocalState
> {
  readonly state: SearchLocalState = initialState;
  private debouncedOnBlur;

  constructor(props) {
    super(props);
    // we need to delay the closing of the popup on search input blur or
    // else the menu items will not fire a click event
    this.debouncedOnBlur = debounce(this.onSearchBlur, 200);
  }

  render() {
    const isPopupOpen =
      this.state.searchHasFocus &&
      this.props.search &&
      this.props.search.trim().length > 0;
    return (
      <div style={style(containerStyle)}>
        <SearchInput
          value={this.props.search}
          onChange={this.onSearch}
          onClear={this.onClear}
          onFocus={this.onSearchFocus}
          onBlur={this.debouncedOnBlur}
          onKeyDown={this.onKeyboard}
          style={style(searchInputStyle)}
          placeholder="search authors, files or commits"
        />
        <Popup isOpen={isPopupOpen} style={style(popupStyle)} noBackdrop>
          {this.renderPopupSuggestions()}
        </Popup>
      </div>
    );
  }

  private renderPopupSuggestions(): JSX.Element[] {
    const unPrefixedSearch = this.getUnprefixedSearch();
    const menuItems: JSX.Element[] = [
      <Selectable
        value="all"
        key="searchSuggestion_all"
        onClick={this.onSuggestClick}
        selected={this.state.selectedSuggestion === 'all'}
      >
        <div>Search everywhere for '{unPrefixedSearch}'</div>
      </Selectable>,
    ];

    return menuItems.concat(
      suggestedSearchPrefixes.map((prefix, index) => {
        return (
          <Selectable
            key={`searchSuggestion_${index}`}
            value={prefix}
            onClick={this.onSuggestClick}
            selected={this.state.selectedSuggestion === prefix}
          >
            <div style={style('normalText')}>
              {prefix} {unPrefixedSearch}
            </div>
          </Selectable>
        );
      })
    );
  }

  private onSearch = value => {
    this.props.dispatch(setSearch(value || ''));
  };

  private onClear = () => {
    this.props.dispatch(setSearch(''));
  };

  private onSearchFocus = () => {
    this.setState({ searchHasFocus: true });
  };

  private onSearchBlur = () => {
    this.setState({ searchHasFocus: false });
  };

  private onSuggestClick = (_evt, value) => {
    this.selectSuggestion(value);
  };

  private selectSuggestion(value) {
    const newSearch =
      value === 'all'
        ? this.getUnprefixedSearch()
        : `${value}${this.getUnprefixedSearch()}`;
    this.props.dispatch(setSearch(newSearch));
    this.setState({ selectedSuggestion: value });
  }

  private onKeyboard = evt => {
    const currentPrefixIndex =
      this.state.selectedSuggestion === 'all'
        ? -1
        : suggestedSearchPrefixes.indexOf(this.state.selectedSuggestion);

    if (evt.key === 'ArrowDown') {
      evt.preventDefault();
      const newSelection =
        currentPrefixIndex < suggestedSearchPrefixes.length - 1
          ? suggestedSearchPrefixes[currentPrefixIndex + 1]
          : 'all';
      this.setState({ selectedSuggestion: newSelection });
    } else if (evt.key === 'ArrowUp') {
      evt.preventDefault();
      const newSelection =
        currentPrefixIndex === 0
          ? 'all'
          : currentPrefixIndex > 0
            ? suggestedSearchPrefixes[currentPrefixIndex - 1]
            : suggestedSearchPrefixes[suggestedSearchPrefixes.length - 1];
      this.setState({ selectedSuggestion: newSelection });
    } else if (evt.key === 'Enter') {
      this.selectSuggestion(this.state.selectedSuggestion);
      return;
    }
  };

  private getUnprefixedSearch() {
    const { search } = this.props;
    const matches = search && search.match(/^[^\:]*\:(.*)/);
    if (!matches || matches.length < 2) {
      return search;
    }
    return matches[1];
  }
}

export default connect(getSearchContainerState)(Search);
