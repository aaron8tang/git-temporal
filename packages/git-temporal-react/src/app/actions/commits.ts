import { ActionTypes } from 'app/actions/ActionTypes';
import { isVscode, vscode } from 'app/actions/vscode';
import { debug } from '@git-temporal/logger';

export const requestCommits = path => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_COMMITS,
});

export const receiveCommits = (path, json) => ({
  selectedPath: path,
  commits: json.commits,
  isFileSelected: json.isFile,
  type: ActionTypes.RECEIVE_COMMITS,
});

const fetchCommits = path => dispatch => {
  dispatch(requestCommits(path));
  if (isVscode) {
    debug('sending history request to vscode ', path);
    // see actions/vscode.ts for response handling that comes as a window event
    vscode.postMessage({ path, command: 'history' });
  } else {
    const pathParam = path && path.trim().length > 0 ? `?path=${path}` : '';
    // TODO : replace this with serviceBaseUrl when it is in
    fetch(`http://localhost:11966/git-temporal/history${pathParam}`)
      .then(response => response.json())
      .then(json => dispatch(receiveCommits(path, json)));
  }
};

const shouldFetchCommits = (state, path) => {
  if (state.isFetching) {
    return false;
  }
  return (
    !state.commits || state.commits.length <= 0 || state.selectedPath !== path
  );
};

export const fetchCommitsIfNeeded = path => (dispatch, getState) => {
  if (shouldFetchCommits(getState(), path)) {
    return dispatch(fetchCommits(path));
  }
};
