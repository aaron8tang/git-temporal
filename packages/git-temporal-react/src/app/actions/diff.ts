import { ICommit } from 'app/interfaces';
import { dateFilteredCommits } from 'app/selectors/dates';
import { ActionTypes } from 'app/actions/ActionTypes';
import { isVscode, vscode } from 'app/actions/vscode';
import { debug } from '@git-temporal/logger';
import { start } from 'repl';

const requestDiff = (path: string) => ({
  selectedPath: path,
  type: ActionTypes.REQUEST_DIFF,
});

export const receiveDiff = (path: string, diff: any) => ({
  diff,
  selectedPath: path,
  type: ActionTypes.RECEIVE_DIFF,
});

export const setDiffStartCommit = (commitId: string) => ({
  commitId,
  type: ActionTypes.SET_DIFF_START_COMMIT,
});

export const setDiffEndCommit = (commitId: string) => ({
  commitId,
  type: ActionTypes.SET_DIFF_END_COMMIT,
});

export const fetchDiff = (
  path: string,
  commits: ICommit[],
  startDate: number,
  endDate: number
) => (dispatch: any): void => {
  dispatch(requestDiff(path));
  const filteredCommits = dateFilteredCommits(commits, startDate, endDate);
  if (!filteredCommits || filteredCommits.length === 0) {
    return;
  }
  // commits are in descending order
  const rightCommit = filteredCommits[0];
  const leftCommit = filteredCommits.slice(-1)[0];

  if (isVscode) {
    debug('sending diff request to vscode ', path, leftCommit, rightCommit);
    // see actions/vscode.ts for response handling that comes as a window event
    vscode.postMessage({
      path,
      leftCommit: leftCommit.id,
      rightCommit: rightCommit.id,
      command: 'diff',
    });
  } else {
    const pathParam =
      path && path.trim().length > 0 ? `?path=${path}` : '?path=.';
    const leftCommitParam = leftCommit ? `&leftCommit=${leftCommit.id}` : '';
    const rightCommitParam = rightCommit
      ? `&rightCommit=${rightCommit.id}`
      : '';

    // TODO : replace this with serviceBaseUrl when it is in
    const url = `http://localhost:11966/git-temporal/diff${pathParam}${leftCommitParam}${rightCommitParam}`;
    fetch(url)
      .then(response => response.json())
      .then(diff => dispatch(receiveDiff(path, diff)));
  }
};
