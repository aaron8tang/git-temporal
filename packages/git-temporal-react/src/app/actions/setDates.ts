import { debug } from '@git-temporal/logger';

import { setStartDate, setEndDate } from 'app/actions';
import { fetchDiff } from 'app/actions/diff';
import { ICommit } from 'app/interfaces';

export function setDates(
  dispatch: (args: any) => void,
  commits: ICommit[],
  selectedPath: string,
  startDate: number,
  endDate: number | Date | null
) {
  const epochStartDate = Math.floor(startDate / 1000);
  const epochEndDate = endDate && Math.floor((endDate as number) / 1000);

  debug('Timeplot: setDates', startDate, epochStartDate, endDate, epochEndDate);
  const [newStartDate, newEndDate] =
    epochStartDate === epochEndDate
      ? [epochStartDate, null]
      : !epochEndDate || epochStartDate < epochEndDate
        ? [epochStartDate, epochEndDate]
        : [epochEndDate, epochStartDate];

  dispatch(setStartDate(newStartDate));
  dispatch(setEndDate(newEndDate));
  dispatch(fetchDiff(selectedPath, commits, newStartDate, newEndDate));
}
