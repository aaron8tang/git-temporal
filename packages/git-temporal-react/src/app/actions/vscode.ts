import { receiveCommits, receiveCommitRange } from 'app/actions/commits';
import { receiveDiff } from 'app/actions/diff';
import { debug } from 'app/utilities/logger';

// @ts-ignore
export const isVscode = window && window.IS_VSCODE_WEBVIEW;
// @ts-ignore
export const vscode = isVscode ? acquireVsCodeApi() : null;

if (isVscode) {
  debug('running in VSCode.');
}

export function handleVscodeMessages(dispatch) {
  // @ts-ignore
  if (!window || !window.IS_VSCODE_WEBVIEW) {
    return;
  }
  window.addEventListener('message', event => {
    const { data } = event;
    debug(`actions/vscode received window message`, {
      ...data,
      // don't dump full commits or file contents to log
      commits: data.commits ? `object[${data.commits.length}]` : undefined,
      leftFileContents: redactArray(data.leftFileContents),
      rightFileContents: redactArray(data.rightFileContents),
      modifiedFiles: redactArray(data.modifiedFiles),
    });

    switch (data.type) {
      case 'commitRange':
        dispatch(receiveCommitRange(data.path, data));
        break;
      case 'history':
        dispatch(receiveCommits(data.path, data));
        break;
      case 'diff':
        dispatch(receiveDiff(data.path, data));
        break;
    }
  });
}

function redactArray(array) {
  return array ? `[..] length=${array.length}` : undefined;
}
