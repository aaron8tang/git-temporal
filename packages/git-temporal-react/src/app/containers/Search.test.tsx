import React from 'react';
import { shallow } from 'enzyme';
import { selectPath } from 'app/actions';

import { Search } from './Search';

describe('containers/Search', () => {
  describe('when rendered without props', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      // @ts-ignore
      selectPath.mockClear();
      mockDispatch = jest.fn();
      wrapper = shallow(<Search dispatch={mockDispatch} search="" />);
    });

    test('it should match snapshot (should render search without popup)', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
  describe('when rendered with search text', () => {
    let wrapper;
    let mockDispatch;
    beforeAll(() => {
      // @ts-ignore
      selectPath.mockClear();
      mockDispatch = jest.fn();
      wrapper = shallow(<Search dispatch={mockDispatch} search="someone" />);
    });

    test('it should match snapshot (it should have rendered search and the popup', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
