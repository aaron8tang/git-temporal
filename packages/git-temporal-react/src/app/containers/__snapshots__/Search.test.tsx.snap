// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Search when rendered with search text and then on focus, it should show popup 1`] = `
<div
  style={
    Object {
      "float": "right",
      "overflow": "visible",
      "position": "relative",
      "width": 250,
    }
  }
>
  <SearchInput
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="search authors, files or commits"
    style={
      Object {
        "borderRadius": 10,
        "marginLeft": 10,
        "marginTop": 0,
        "width": "100%",
      }
    }
    value="someone"
  />
  <Popup
    isOpen={true}
    noBackdrop={true}
    style={
      Object {
        "marginRight": 0,
      }
    }
  >
    <Selectable
      key="searchSuggestion_all"
      onClick={[Function]}
      selected={true}
      value="all"
    >
      <div>
        Search everywhere for '
        someone
        '
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_0"
      onClick={[Function]}
      selected={false}
      value="author:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        author:
         
        someone
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_1"
      onClick={[Function]}
      selected={false}
      value="commit:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        commit:
         
        someone
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_2"
      onClick={[Function]}
      selected={false}
      value="file:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        file:
         
        someone
      </div>
    </Selectable>
  </Popup>
</div>
`;

exports[`containers/Search when rendered with search text it should match snapshot (it should have rendered search, but NOT the popup 1`] = `
<div
  style={
    Object {
      "float": "right",
      "overflow": "visible",
      "position": "relative",
      "width": 250,
    }
  }
>
  <SearchInput
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="search authors, files or commits"
    style={
      Object {
        "borderRadius": 10,
        "marginLeft": 10,
        "marginTop": 0,
        "width": "100%",
      }
    }
    value="someone"
  />
  <Popup
    isOpen={false}
    noBackdrop={true}
    style={
      Object {
        "marginRight": 0,
      }
    }
  >
    <Selectable
      key="searchSuggestion_all"
      onClick={[Function]}
      selected={true}
      value="all"
    >
      <div>
        Search everywhere for '
        someone
        '
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_0"
      onClick={[Function]}
      selected={false}
      value="author:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        author:
         
        someone
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_1"
      onClick={[Function]}
      selected={false}
      value="commit:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        commit:
         
        someone
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_2"
      onClick={[Function]}
      selected={false}
      value="file:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        file:
         
        someone
      </div>
    </Selectable>
  </Popup>
</div>
`;

exports[`containers/Search when rendered without props it should match snapshot (should render search without popup) 1`] = `
<div
  style={
    Object {
      "float": "right",
      "overflow": "visible",
      "position": "relative",
      "width": 250,
    }
  }
>
  <SearchInput
    onBlur={[Function]}
    onChange={[Function]}
    onClear={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    placeholder="search authors, files or commits"
    style={
      Object {
        "borderRadius": 10,
        "marginLeft": 10,
        "marginTop": 0,
        "width": "100%",
      }
    }
    value=""
  />
  <Popup
    isOpen={false}
    noBackdrop={true}
    style={
      Object {
        "marginRight": 0,
      }
    }
  >
    <Selectable
      key="searchSuggestion_all"
      onClick={[Function]}
      selected={true}
      value="all"
    >
      <div>
        Search everywhere for '
        '
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_0"
      onClick={[Function]}
      selected={false}
      value="author:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        author:
         
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_1"
      onClick={[Function]}
      selected={false}
      value="commit:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        commit:
         
      </div>
    </Selectable>
    <Selectable
      key="searchSuggestion_2"
      onClick={[Function]}
      selected={false}
      value="file:"
    >
      <div
        style={
          Object {
            "color": "#333333",
            "fontSize": 13,
            "fontWeight": "normal",
          }
        }
      >
        file:
         
      </div>
    </Selectable>
  </Popup>
</div>
`;
